proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,618 @@
+
+The following expressions are parse errors:
+
+* ``s"a ${} b"``
+
+  * Expression is missing
+
+* ``s"a ${=} b"``
+
+  * Not an expression
+
+* ``s"a ${let x =} b"``
+
+  * Incomplete expression
+
+* ``s"a ${{b} c"``
+
+  * The second ``{`` is not a valid character to start an expression
+
+* ``s"a ${'}'} b"``
+
+  * This would try to lex the expression ``'``
+
+* ``s"foo ${x {- inline comment -} } bar"``
+
+  * This would try to lex the expression ``x {- inline comment -``
+
+* ``s"foo ${User{id = 123}} bar"``
+
+  * This would try to lex the expression ``User{id = 123``
+
+See the "Semantics" and "Examples" sections to see examples of valid interpolated strings.
+
+Machinery
+~~~~~~~~~
+
+The machinery should satisfy the following requirements:
+
+* Allows constructing any type, not just String or String-like types
+
+  * See the "SqlQuery" or "HTML" examples
+
+* Allows interpolating any type, not just String
+
+  * Interpolate Showable things with show by default
+  * Allow overriding interpolation for specific types
+
+An interpolated string expression desugars to calls to ``fromBuilder``, ``toBuilder``, and ``interpolate``, which are defined as:
+
+::
+
+  -- | Laws:
+  --     * fromBuilder . toBuilder === id
+  class Monoid (Builder s) => Buildable s where
+    type Builder s = r | r -> s
+    toBuilder :: s -> Builder s
+    fromBuilder :: Builder s -> s
+
+  class Buildable s => Interpolate a s where
+    interpolate :: a -> Builder s
+
+Instances for ``String`` will be defined as well:
+
+::
+
+  newtype StringBuilder = StringBuilder (Endo String)
+    deriving newtype (Semigroup, Monoid)
+
+  instance Buildable String where
+    type Builder String = StringBuilder
+    toBuilder s = StringBuilder (Endo (s ++))
+    fromBuilder (StringBuilder (Endo f)) = f []
+
+  instance Interpolate String String where
+    interpolate = toBuilder
+  instance Interpolate Char String where
+    interpolate = interpolate . (:[])
+  instance {-# OVERLAPPABLE #-} Show a => Interpolate a String where
+    interpolate = StringBuilder . Endo . shows
+
+This design allows interpolating anything, even user-defined types, into a ``String`` with ``Show``, but can be overridden for specific types. See the "Examples" section for more details.
+
+These definitions would initially be implemented in ``ghc-experimental`` under ``Data.String.Interpolate``. After the API has stablized, these would eventually live in ``GHC.Exts`` alongside ``IsString``.
+
+Expansion
+~~~~~~~~~
+
+With the machinery defined above, the following interpolated string desugars to the below expression:
+
+::
+
+  -- original string
+  s"foo ${f a b} bar ${g x} baz ${name}"
+
+  -- desugared
+  fromBuilder $
+    toBuilder "foo "
+    <> interpolate (f a b) <> toBuilder " bar "
+    <> interpolate (g x)   <> toBuilder " baz "
+    <> interpolate name    <> toBuilder ""
```

Comment:
Is the implication that there will always be literal strings surrounding any interpolation? Does that mean that ```"${x}${y}"``` becomes ```toBuilder "" <> interpolate x <> toBuilder "" <> interpolate y <> toBuilder ""```?

Reply:
That was the original design, yes, but there might be changes here

Reply:
Note that for strings, there can be no semantic difference between `fromBuilder (toBuilder "" <> interpolate x)` and `fromBuilder (interpolate x)` due to the laws associated with `Buildable`:

```
fromBuilder (toBuilder "" <> interpolate x)
=== { monoid homo }
fromBuilder (toBuilder "") <> fromBuilder (interpolate x)
=== { embedding }
"" <> fromBuilder (interpolate x)
=== { "" is neutral element of <> }
fromBuilder (interpolate x)
```

So the exact specification does not matter for `String`.

It matters for other `Interpolate` instances, though.. (about to take reference to this reply in https://github.com/ghc-proposals/ghc-proposals/pull/570#discussion_r1713033502).